src/plugins/kibana_usage_collection/server/collectors/cloud/detector/cloud_service.ts

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import fs from 'fs';
+import { isObject, isString } from 'lodash';
+import defaultRequest from 'request';
+import type { OptionsWithUri, Response as DefaultResponse } from 'request';
+import { CloudServiceResponse } from './cloud_response';
+
+/** @internal */
+export type Request = typeof defaultRequest;
+
+/** @internal */
+export type RequestOptions = OptionsWithUri;
+
+/** @internal */
+export type Response = DefaultResponse;
+
+/** @internal */
+export interface CloudServiceOptions {
+  _request?: Request;
+  _fs?: typeof fs;
+  _isWindows?: boolean;
+}
+
+/**
+ * CloudService provides a mechanism for cloud services to be checked for
+ * metadata that may help to determine the best defaults and priorities.
+ */
+export class CloudService {
+  private readonly _request: Request;
+  protected readonly _name: string;
+
+  constructor(name: string, options: CloudServiceOptions = {}) {
+    this._name = name.toLowerCase();
+
+    // Allow the HTTP handler to be swapped out for tests
+    const { _request = defaultRequest } = options;
+
+    this._request = _request;
+  }
+
+  /**
+   * Get the search-friendly name of the Cloud Service.
+   */
+  getName() {
+    return this._name;
+  }
+
+  /**
+   * Using whatever mechanism is required by the current Cloud Service,
+   * determine if Kibana is running in it and return relevant metadata.
+   */
+  checkIfService() {
+    return this._checkIfService(this._request).catch(() => this._createUnconfirmedResponse());
+  }
+
+  _checkIfService(request?: Request): Promise<CloudServiceResponse> {
+    // should always be overridden by a subclass
+    return Promise.reject(new Error('not implemented'));
+  }
```

Comment:
Super-NIT: Maybe we can declare this method (or the whole class) as `abstract` so we don't need this type of implementations since we won't be allowed to use it at all. What do you think?